[스프링4기 2주차] Spring Web 으로 ToDo REST API 만들기

hyunaseodev opened the pull request.

hyunaseodev committed:
수정 버튼 생성

hyunaseodev committed:
Spring을 활용하여 GET/tasks, POST/tasks 구현

hyunaseodev committed:
Spring을 활용하여 PATCH/task 구현

한윤석님의 지원과 도움을 받음.

hyunaseodev committed:
Spring을 활용하여 GET /tasks/{id}, DELETE /tasks 구현

추가로 PATCH /tasks에서 예외발생시 해결하는 코드 작성

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,244 @@
+        // 찾을 수 없는 경우 -1 반환. IndexOutOfBoundsException 오류 발생.
+        // IndexOutOfBoundsException 오류 해결하기 위해 if문 사용.
+        // if문 조건식인 index 값이 -1이 아닌 경우 문장 수행
+        Task foundTask = null;
+        if(taskIndex != -1){
+            tasks.get(taskIndex);
+            // * 찾은 Task를 수정해요.
+            foundTask.setTitle(task.getTitle());
+        }
+        // * 수정된 Task를 응답한다.
+        return foundTask;
```
찾을 수 없을 경우 `null`을 반환하여 아무것도 반환하지 않는다 라고 결정하신 것 같아요. 그래서 foundTask를 null로 초기화를 해놓고, 할일을 찾은 경우에만 foundTask의 값을 갱신하여 title을 수정하려고 하신 것 같습니다.

 그런데 지금 null로 초기화를 하고 찾았을 경우 foundTask에 다시 재할당을 하지 않아 Null point exception이 발생할 것 같아요. 따라서 다음과 같이 해야할 것 같습니다.
```suggestion
        Task foundTask = null;
        if(taskIndex != -1){
            foundTask = tasks.get(taskIndex);
            foundTask.setTitle(task.getTitle());
        }

        return foundTask;
```

위의 방법도 좋은 방법이지만, foundTask가 처음에 null로 초기화되고 다시 재할당하는 방법은 변수가 언제 어디서 바뀔지 예측하기 어렵다는 단점이 있습니다. 지금같은 경우는 코드가 7줄 정도밖에 되지 않아서 이해할 수 있지만, 나중에 중간에 다른 복잡한 로직이 껴지면 `변수가 변할 수 있다` 따라서 코드를 읽을 때 기억하며 읽어야 하기 때문에 복잡성을 증가시킵니다. 따라서 찾을 수 없는 경우 빠르게 로직을 종료하여 더 이상 코드를 읽을 필요 없게하고, 변수가 변하지 않는 상수이기 때문에 변하는 것에 대한 두려움 없이 코드를 읽을 수 있도록 할 수 있습니다.

```java
if(taskIndex < 0) {
    return null;
}

Task foundTask = tasks.get(taskIndex);
foundTask.setTitle(task.getTitle());

return foundTask;
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,244 @@
+            파라미터를 받는 방법은 @PathVariable 와 @RequestParam
+            둘의 차이를 자세히 알아본 후 어느 방법이 더 효율적인지 정한 후 적용하기.
+    */
+    @GetMapping("/{paramId}")
```
받는 것이 Task의 id니까 `taskId`라고 쓰는 것은 어떨까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,244 @@
+    3. 실행
+    4. 회고
+        - 어떻게 하면 더 효과적으로 해결할 수 있었을까?
+            for문을 사용하여 할 일 목록을 찾는 경우 오래걸릴 가능성 있음. 속도개선을 위해 다른 방법 고민 및 활용하기.
```
성능에 대해서 고민을 하고 계시군요 ㅎㅎ 할 일을 List가 아닌 HashMap으로 관리하는 것도 좋을 것 같아요

hannut91 commented:
풀리퀘스트 내용에 어떤 일을 했는지 자세히 적어주시면 같이 일하는 팀원이 파악하기 더 좋을 것 같습니다

hyunaseodev committed:
2주차 과제 코멘트 적용

1. 변수명 변경
paramId -> taskId
2. 찾을 수 없을 경우 
foundTask 를 재할당하는 방법 대신 빠르게 로직 종료하는 방법 적용.
(더 이상 코드를 읽을 때 기억할 필요 없으며 변수가 변하지 않는 상수이므로 두려움과 복잡성 해소)

hannut91 commented:
이번 한 주도 고생 많으셨습니다! 주간회고를 작성해 주세요. 회고를 작성하면서 지난 한 주를 돌아보고 좋았던 점은 다음 주 과제를 진행할 때 더 개선해보고, 아쉬웠던 점은 어떻게 보완하면 좋을 것인지를 중심적으로 적어주시면 좋을 것 같아요

See also
  - https://dal-lab.com/2019/09/18/today-i-learned/
